pkg/parquetquery/predicates.go

```diff
@@ -485,3 +434,47 @@ func (p *InstrumentedPredicate) KeepValue(v pq.Value) bool {
 
 	return false
 }
+
+// DictionaryPredicateHelper is a helper for a predicate that uses a dictionary
+// for filtering.
+//
+// There is one dictionary per ColumnChunk/RowGroup, but it is not accessible in
+// KeepColumnChunk. This helper saves the result of KeepPage and uses it for
+// all pages in the row group. It also has a basic heuristic for choosing not
+// to check the dictionary at all if the cardinality is too high.
+type DictionaryPredicateHelper struct {
+	newRowGroup         bool
+	keepPagesInRowGroup bool
+}
+
+func (d *DictionaryPredicateHelper) setNewRowGroup() {
+	d.newRowGroup = true
+}
+
+func (d *DictionaryPredicateHelper) keepPage(page pq.Page, keepValue func(pq.Value) bool) bool {
+	if !d.newRowGroup {
+		return d.keepPagesInRowGroup
+	}
+
+	d.newRowGroup = false
+	d.keepPagesInRowGroup = true
+
+	// If a dictionary column then ensure at least one matching
+	// value exists in the dictionary
+	dict := page.Dictionary()
+	if dict == nil {
+		return d.keepPagesInRowGroup
+	}
```

Comment:
`nitpick` Moving the assignment  into the `dict == nil` condition makes the intent a bit clearer
```suggestion
	d.newRowGroup = false

	// If a dictionary column then ensure at least one matching
	// value exists in the dictionary
	dict := page.Dictionary()
	if dict == nil {
		d.keepPagesInRowGroup = true
		return d.keepPagesInRowGroup
	}
```

Reply:
I appreciate the comment, but I actually like it where it is. B/c this is a new row group lines 459-460 are resetting to the "default state" which is keepPagesInRowGroup = true. If we were to add a new conditions before the dictionary loop we would want to inherit this default state.